### PR TITLE
BUG: fix an incoming incompatibility with numpy 2.1 in `utils.masked`

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
xref: https://github.com/numpy/numpy/pull/26846

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Ah, it's failing the devdeps jobs because numpy nightlies have not yet caught up with the change this is addressing. For reference, I discovered and fixed it locally by building numpy's main branch

**Comment by neutrinoceros** [comment]
Technically ready for review so I’ll undraft now, but maybe others would rather wait for numpy nightlies to catch up before this goes in.

**Review by mhvk — APPROVED** [review]
Yes, when I reviewed the numpy code changes, I feared we had to make changes too... But this all looks good. But let's wait with merging until the nightlies have updated. Thanks!

**Comment by mhvk** [comment]
The code was basically a copy from what was in `arrayprint.py`, with just a comment stating what changed. I think it still makes sense to do that, but possibly that code has in the meantime changed too...  In any case, my sense would be to stick with what you have, but I'm also happy with moving to `getprintoptions`...

**Comment by neutrinoceros** [comment]
I gave it a try and found something interesting:

here are the options we get with the current implementation without altering defaults:
```
               {'edgeitems': 3,
                'floatmode': 'maxprec',
                'formatter': None,
                'infstr': 'inf',
                'legacy': 9223372036854775807,
                'linewidth': 75,
                'nanstr': 'nan',
                'override_repr': None,
                'precision': 8,
                'sign': '-',
                'suppress': False,
                'threshold': 1000}
```
and here's what we get instead from `np.get_printoptions`
```
             {'edgeitems': 3,
              'floatmode': 'maxprec',
              'formatter': None,
              'infstr': 'inf',
              'legacy': False,
              'linewidth': 75,
              'nanstr': 'nan',
              'override_repr': None,
              'precision': 8,
              'sign': '-',
              'suppress': False,
              'threshold': 1000}
```

notice the value of `legacy` changed (and unfortunately, this *does* have an impact on some of our tests). From looking at `np.get_printoptions`'s implementation, this is very clearly an intentional behavior there, so I *think* it may be considered a (minor) bug on our side that we apparently rely on implementation details from numpy for some tests, but I think fixing this can wait for astropy 7.0, so I'll open another PR to propose this change.

**Comment by neutrinoceros** [comment]
done in  #16759

**Comment by pllim** [comment]
tl;dr -- We need another commit here to fix this failure in devdeps?

`ModuleNotFoundError: No module named 'numpy._core.printoptions'`

**Comment by pllim** [comment]
OK nvm I just saw https://github.com/astropy/astropy/pull/16755#issuecomment-2244454274

I will come back to this when the "nightly" updates, I guess. Thanks!

**Comment by neutrinoceros** [comment]
If I'm reading the cron right, the next batch of nightlies is expected in about 6 hours, so I'll rebase this tomorrow morning and hopefully CI should go full green here.

**Comment by neutrinoceros** [comment]
- nightlies are caught up
- and devdeps is now passing
- but windows is down (#16765)

(semi-accidental haiku)

**Comment by neutrinoceros** [comment]
Incidentally, if you want to see tests failing *without* this patch, see [example logs](https://github.com/astropy/astropy/actions/runs/10072439171/job/27844396122)

**Comment by pllim** [comment]
Thanks! Windows failure is indeed unrelated.

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/issues/16851#issuecomment-2296662194

Fixed the milestone.
